File: generated_code/cpp/solution_2903_3.cpp
#include <numeric>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 * int val;
 * ListNode *next;
 * ListNode() : val(0), next(nullptr) {}
 * ListNode(int x) : val(x), next(nullptr) {}
 * ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        // If the list is empty or has only one node, no adjacent pairs exist.
        if (!head || !head->next) {
            return head;
        }
        
        ListNode* curr = head;
        
        // Traverse the list until we reach the last node
        while (curr && curr->next) {
            // Calculate the GCD of the current node and the next node
            int gcd_val = std::gcd(curr->val, curr->next->val);
            
            // Create a new node with the GCD value, pointing to the next node
            ListNode* gcd_node = new ListNode(gcd_val, curr->next);
            
            // Link the current node to the newly created GCD node
            curr->next = gcd_node;
            
            // Move curr forward by two steps to point to the original next node
            curr = gcd_node->next;
        }
        
        return head;
    }
};

File: generated_code/python/solution_2903_3.py
import math
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Edge case: if the list is empty or has only one node, no insertions are needed.
        if not head or not head.next:
            return head
        
        # Use a single pointer to traverse the list and check adjacent pairs.
        # Time Complexity: O(N * log(min(A, B))) where N is the number of nodes.
        # Space Complexity: O(1) auxiliary space as we modify the list in place.
        curr = head
        while curr and curr.next:
            # Calculate the greatest common divisor of the current and next node values
            gcd_val = math.gcd(curr.val, curr.next.val)
            
            # Create the new node with the GCD value
            gcd_node = ListNode(val=gcd_val)
            
            # Insert gcd_node between curr and curr.next
            gcd_node.next = curr.next
            curr.next = gcd_node
            
            # Move the pointer past the newly inserted node to the original next node
            curr = gcd_node.next
            
        return head